src/V3SplitVar.cpp

```diff
@@ -0,0 +1,1258 @@
+    return NULL;
+}
+
+static const char* cannotSplitConnectedPortReason(AstPin* pinp) {
+    AstVar* varp = pinp->modVarp();
+    if (!varp) return "it is not connected";
+    if (const char* reason = cannotSplitVarDirectionReason(varp->direction())) return reason;
+    return NULL;
+}
+
+static const char* cannotSplitTaskReason(const AstNodeFTask* taskp) {
+    if (taskp->prototype()) return "the task is prototype declaration";
+    if (taskp->dpiImport()) return "the task is imported from DPI-C";
+    if (taskp->dpiOpenChild()) return "the task takes DPI-C open array";
+    return NULL;
+}
+
+static const char* cannotSplitVarCommonReason(const AstVar* varp) {
+    if (AstNodeFTask* taskp = VN_CAST(varp->backp(), NodeFTask)) {
+        if (const char* reason = cannotSplitTaskReason(taskp)) return reason;
+    }
+    if (const char* reason = cannotSplitVarTypeReason(varp->varType())) return reason;
+    if (const char* reason = cannotSplitVarDirectionReason(varp->direction())) return reason;
+    if (varp->isSigPublic()) return "it is public";
+    if (varp->isInoutish()) return "it is bidirectional";
+    if (varp->isUsedLoopIdx()) return "it is used as a loop variable";
+    if (varp->isGenVar()) return "it is genvar";
+    if (varp->isParam()) return "it is parameter";
+    return NULL;
+}
+
+static const char* cannotSplitPackedVarReason(const AstVar* varp);
+
+template <class ALWAYSLIKE>
+void InsertBeginCore(ALWAYSLIKE* ap, AstNodeStmt* stmtp, AstNodeModule* modp) {
+    if (ap->isJustOneBodyStmt() && ap->bodysp() == stmtp) {
+        stmtp->unlinkFrBack();
+        // Insert begin-end because temp value may be inserted to this block later.
+        const std::string name = "__SplitVar__Blk" + cvtToStr(modp->varNumGetInc());
+        ap->addStmtp(new AstBegin(ap->fileline(), name, stmtp));
+    }
+}
+
+void InsertBeginCore(AstInitial* initp, AstNodeStmt* stmtp, AstNodeModule* modp) {
+    if (initp->isJustOneBodyStmt() && initp->bodysp() == stmtp) {
+        stmtp->unlinkFrBack();
+        // Insert begin-end because temp value may be inserted to this block later.
+        const std::string name = "__SplitVar__Blk" + cvtToStr(modp->varNumGetInc());
+        initp->replaceWith(
+            new AstInitial(initp->fileline(), new AstBegin(initp->fileline(), name, stmtp)));
+        VL_DO_DANGLING(initp->deleteTree(), initp);
+    }
+}
+
+static void InsertBeginIfNecessary(AstNodeStmt* stmtp, AstNodeModule* modp) {
+    AstNode* backp = stmtp->backp();
+    if (AstAlways* ap = VN_CAST(backp, Always)) {
+        InsertBeginCore(ap, stmtp, modp);
+    } else if (AstAlwaysPublic* ap = VN_CAST(backp, AlwaysPublic)) {
+        InsertBeginCore(ap, stmtp, modp);
+    } else if (AstInitial* ap = VN_CAST(backp, Initial)) {
+        InsertBeginCore(ap, stmtp, modp);
+    }
+}
+
+// Assume startp is a port variable, traverse until the end of port declaration.
+// Returned pointer is the end of port variable.
+static AstNode* toEndOfPort(AstVar* startp) {
```

Comment:
Why can't you always insert right after the given port?

This adds a potential O(ports^2) iteration.

Reply:
Well, I don't remember well, but I guess AST looked clean and friendly to me.
Now the function is removed.